Implement "IsSupported" for all ISA classes of Intel hardware intrinsics

fiigii opened the pull request:
This PR is implementing  `IsSupported` properties for all ISA classes in `System.Runtime.Intrinsics.X86` under JIT compilation (not work with AOT).

Close #13982

fiigii commented:
cc @russellhadley @jkotas @AndyAyersMS @CarolEidt @BruceForstall

jkotas commented on src/mscorlib/shared/System/Runtime/CompilerServices/IntrinsicAttribute.cs:
```diff
@@ -6,7 +6,7 @@ namespace System.Runtime.CompilerServices
     // Calls to methods or references to fields marked with this attribute may be replaced at
     // some call sites with jit intrinsic expansions.
-    [AttributeUsage(AttributeTargets.Method | AttributeTargets.Constructor | AttributeTargets.Field, Inherited = false)]
+    [AttributeUsage(AttributeTargets.Method | AttributeTargets.Constructor | AttributeTargets.Field | AttributeTargets.Property, Inherited = false)]
```
This does not make sense on property. It has to be on the method implementing the property, like:

```
public static bool IsSupported
{
    [Intrinsic]
    get { return false; }
}
```

fiigii replied:
Will fix.

jkotas commented on src/mscorlib/src/System/Runtime/Intrinsics/X86/Aes.cs:
```diff
@@ -13,6 +14,7 @@ namespace System.Runtime.Intrinsics.X86
     [CLSCompliant(false)]
     public static class Aes 
     {
+        [Intrinsic]
```
Since every method in this namespace will be marked as intrinsic eventually, I am wondering whether we should rather special case `System.Runtime.Intrinsics.*` (here)[https://github.com/dotnet/coreclr/blob/master/src/vm/methodtablebuilder.cpp#L5086] and set the flag for all methods under this namespaces.


https://github.com/dotnet/coreclr/blob/master/src/vm/methodtablebuilder.cpp#L5088

fiigii replied:
> we should rather special case `System.Runtime.Intrinsics.*` ... and set the flag for all methods under this namespaces.

Doesn't this reduce intrinsic recognition overhead?

jkotas replied:
Well, each `[Intrinsic]` adds size to `System.Private.CoreLib`. It is hard to tell what is faster, and it is not that important because the result is cached. Special casing the namespace will be definitely smaller (rought estimate 5-10 kB of disk footprint).

fiigii replied:
I see. Could we change `IntrinsicsAttribute` to be able to apply on `Class` and `Struct` (or larger scopes)? Then specially treat `System.Runtime.Intrinsics.*`?

jkotas replied:
You cannot apply the attribute on large scopes. It would have to be on each class or struct. It does not look like a particularly great option to me.

fiigii replied:
> It does not look like a particularly great option to me.

How about considering this optimization later (in a new issue)? Or you prefer to address it in this PR?

jkotas replied:
If we do it now, it would save you from editing like dozen of .cs files. But if you prefer to do it separately - it is fine with me.

fiigii replied:
> If we do it now, it would save you from editing like dozen of .cs files. 

Good point 😄. But let's consider the "optional expand" issue at first, that is more important.

mikedn commented on src/jit/compiler.cpp:
```diff
@@ -2560,6 +2560,79 @@ void Compiler::compSetProcessor()
+
+// Instruction set flags fo// Instruction set flags for Intel hardware intrinsics
+#ifdef _TARGET_XARCH_
+    opts.compSupportsAES       = false;
```
Perhaps a single "flags" variable should be used instead of all these variables?

fiigii replied:
Agree, will change to use a 64-bit flag variable.

mikedn commented on src/jit/compiler.h:
```diff
@@ -2995,6 +2995,28 @@ class Compiler
+    InstructionSet lookupHWIntrinsicISA(const char* className);
+    NamedIntrinsic lookupHWIntrinsic(const char* methodName, InstructionSet isa);
+    InstructionSet isaOfHWIntrinsic(NamedIntrinsic intrinsic);
+    GenTreePtr impX86HWIntrinsic(NamedIntrinsic intrinsic, CORINFO_METHOD_HANDLE method);
```
`GenTree*` instead of `GenTreePtr`

fiigii replied:
Thanks, will do. Shall we always use `GenTree*` instead of `GenTreePtr`?

mikedn replied:
I suggested this in other PRs and it seems that there is universal agreement about using `GenTree*` instead of `GenTreePtr`.

There is no `GenTreeCastPtr`, no `BasicBlockPtr`, no `CompilerPtr`. It's just `GenTreePtr`, an oddball.

fiigii replied:
Got it.

mikedn commented on src/jit/importer.cpp:
```diff
@@ -3865,9 +3870,392 @@ NamedIntrinsic Compiler::lookupNamedIntrinsic(CORINFO_METHOD_HANDLE method)
     return result;
 }
 
+#ifdef _TARGET_XARCH_
```
Could all this new code be moved to a separate file? `importerxarch.cpp` perhaps? This file is already large as is.

fiigii replied:
Yes, I will separate this new code from importer.cpp to a new file. How about `hwintrinsicxarch.pp`?

mikedn replied:
> How about hwintrinsicxarch.pp?

I assume you mean .cpp. Sounds reasonable to me.

fiigii replied:
Yes 😄

mikedn commented on src/jit/lsraxarch.cpp:
```diff
@@ -2261,7 +2261,7 @@ void LinearScan::TreeNodeInfoInitSIMD(GenTreeSIMD* simdTree)
                 {
                     bool needFloatTemp;
                     if (varTypeIsSmallInt(simdTree->gtSIMDBaseType) &&
-                        (compiler->getSIMDInstructionSet() == InstructionSet_AVX))
```
Isn't this change part of a different PR?

fiigii replied:
Please see https://github.com/dotnet/coreclr/pull/13879

fiigii commented:
@mellinoe @jkotas Tests failed on
```
The type or namespace name 'Intrinsics' does not exist in the namespace 'System.Runtime' (are you missing an assembly reference?) [D:\j\workspace\debug_windows---17180f1b\tests\src\JIT\HardwareIntrinsics\IsSupported.csproj
```
Should I set the assembly reference somewhere?

mellinoe commented:
@fiigii That is a good question. The test dependencies are defined [here](https://github.com/fiigii/coreclr/blob/3a7c1e3e8aba2711cbdd24367db98b0c8cafd601/tests/src/Common/test_dependencies/test_dependencies.csproj) (I think -- it might be in the test_runtime.csproj next to it)

But currently, System.Runtime.Intrinsics.x86 is actually not being packaged up at all in corefx. So there's nothing for us to reference to get it here. We probably need to create a standalone package containing the assembly. I can work on that.

fiigii commented:
## Update  
- Moved hardware intrinsic related code to a new file "hwintrinsicxarch.cpp".
- Removed `[Intrinsics]` from hardware intrinsic definition, and specially treat `System.Runtime.Intrinsics.X86` namespace in intrinsic recognition.
- Changed ISA variables to a flag variable.
- Hardcoded the version of  `System.Runtime.Intrinsics.X86` package to resolve the building error. (Any better solution?).
- Revoked changes from https://github.com/dotnet/coreclr/pull/13879. I carefully retain the name of `InstructionSet_AVX` in legacy SIMD code. I will address the codegen conflict in #14065 rather than just name conflict.

fiigii commented:
cc @jkotas @mikedn @russellhadley

fiigii commented:
These intrinsics cannot be expanded under debug build. @jkotas , do you have any suggestion to address the "optional expansion" issue?

jkotas commented on src/vm/methodtablebuilder.cpp:
```diff
@@ -5094,6 +5094,19 @@ MethodTableBuilder::InitNewMethodDesc(
+        LPCUTF8 className;
+        LPCUTF8 nameSpace;
+
+        HRESULT hrns = GetMDImport()->GetNameOfTypeDef(bmtInternal->pType->GetTypeDefToken(), &className, &nameSpace);
```
It would be better to do this check once per type, not once per each method.

jkotas replied:
Also, it can be under `defined(_TARGET_X86_) || defined(_TARGET_AMD64_)` ifdef.

fiigii replied:
Thanks, will do.

jkotas commented:
My proposal would be:

- Implement the instrinsics by recursively calling itself, e.g. `public static Vector256<float> Add(Vector256<float> left, Vector256<float> right) => Add(left, right)`.
- When the JIT sees intrinsic recursively calling itself, it should force expand it.

This plan will:
- Allows the JIT to no expand the intrinsics with optimizations off. This is important for both throughput and resiliency against bugs (when there is a fatal error during JITing, the JIT tried to compile the method again with optimizations off).
- Makes the intrinsics work even when the optimizations are off (they will be calls).
- Makes the intrinsics callable by reflection. This is important for parts of the system that interpret the IL.

jkotas commented:
Also, we will need to disable saving of compiled intrinsic method bodies into NGen images. The `IsSupported` intrinsics at least. There needs to be a fresh body generated at runtime based on the actual processor capabilities.

fiigii commented:
> When the JIT sees intrinsic recursively calling itself, it should force expand it.

This solution makes sense to me, but it will complicate the intrinsic implementation for other hardware platforms.  Originally, I planned to let x86 intrinsics back to C# code on ARM, which always returns `false` from `IsSupported` and throws exceptions from others.

> There needs to be a fresh body generated at runtime based on the actual processor capabilities.

Don't you mean that we can compile intrinsics and insert into AOT-compiled images at runtime?

fiigii commented:
> You can still do that by having two files with the intrinsics implementations - one throwing (included in non-xarch CoreLib builds) and one recursive (included in xarch CoreLib builds).

This solution sounds pretty cool 😎 , will do.

> Yes, you can say it that way.

This AOT solution would be perfect if the start time regression is acceptable. I want to implement all the intrinsics in JIT at first and  enable them for AOT later (the AOT work may need your help 😄 ).

CarolEidt commented on src/jit/instr.h:
```diff
@@ -280,10 +280,29 @@ enum emitAttr : unsigned
+    // Linear order end
+    // Instruction sets have the linear order only in above area.
+    // Values of InstructionSet not in this area cannot be compared 
+    // (i.e. compiler->getSIMDInstructionSet() >= InstructionSet_SSE3_4).
```
It appears that the values can no longer be compared at all. Indeed, this change would seem to break the existing code that does comparisons, since if any of the higher bits are set, it will be true that `compiler->getSIMDInstructionSet() >= InstructionSet_AVX2`, which could lead to problems. I don't think you can make this change without changing all the code that depends on the values of `compiler->getSIMDInstructionSet()` being in linear order.

fiigii replied:
> It appears that the values can no longer be compared at all.

I agree, and this is a temporary solution. Actually, only SIMD intrinsic (`System.Numerics.Vectors`) code is comparing instruction set values, and the new hardware intrinsic implementation never depends upon the order. In this PR, SIMD still works well because `getSIMDInstructionSet` only returns SSE2, SSE3_4, or AVX. I will refactor this "ISA comparison" in https://github.com/dotnet/coreclr/issues/14065

CarolEidt replied:
> I agree, and this is a temporary solution.

Before merging this, then, you need to add a comment indicating what the longer term solution is, and how it will interplay with `getFloatingPointInstructionSet()` (and `getSIMDInstructionSet()`).

fiigii replied:
Got it. Will do.

CarolEidt commented on tests/src/JIT/HardwareIntrinsics/IsSupported.cs:
```diff
@@ -0,0 +1,89 @@
+                }
+                if (Aes.IsSupported && Bmi1.IsSupported && Bmi2.IsSupported && Fma.IsSupported && 
+                        Lzcnt.IsSupported && Popcnt.IsSupported && Pclmulqdq.IsSupported)
+                    {
```
Indentation is off here.

fiigii replied:
Thanks.

CarolEidt commented on tests/src/JIT/HardwareIntrinsics/IsSupported.cs:
```diff
@@ -0,0 +1,89 @@
+                    else
+                    {
+                        result = false;
+                    }
```
Since it is difficult to do any real verification of the result here, it might be useful to also add a test that the size of, say, Vector<float> is 128, unless Avx2.IsSupported, in which case it should be 256.

fiigii replied:
Good point! Will do.

fiigii commented:
> It appears that this breaks existing code that checks the current instruction set, and

@CarolEidt No, it does not. Currently, the new hardware intrinsic implementation does not overlap with the existing SIMD code at all. There are indeed some naming and potential codegen conflicts that we discussed in https://github.com/dotnet/coreclr/pull/13879, but do not impact the current codegen. I will solve all the conflicts in https://github.com/dotnet/coreclr/issues/14065 .

> This increases the cost of setting the instruction set. That may not be a significant issue, but as we add intrinsics and increase the string comparisons we are doing to identify them, I'm concerned about the impact of JIT throughput. Could you please measure before & after?

I think we are discussing this in https://github.com/dotnet/corefx/issues/23519

AndyAyersMS commented:
It would be nice to expand the tabular scheme you have introduced for the hardware intrinsics to include the non-hardware ones too (right now there's just the one, but there are two more in pending PRs, and likely more to come). 

In my opinion this can wait for a subsequent change where we perhaps also auto-generate the recognizer. But we should do this before opening the floodgates on the hardware intrinsic set.

I wonder what the best scheme is for handling prejitted code. 

The basic option is to have the target-relevant `IsSupported` checks to remain calls and skip prejitting for those methods. I don't know if we're going to allow mixed IAS target intrinsics in method bodies (seems like we could but might not want to) -- if so then when prejitting the off-target `IsSupported`s must evaluate to false.

We could go beyond that as follows.

We could fail prejitting (or the equivalent) when one of these intrinsics is seen in the importer either as a root method or a callee. Hence any method that contains a target-appropriate `IsSupported` call will also get jitted and potentially be tailored to the target ISA. This has the drawback that if A calls B and B is an inline candidate with an `IsSupported` call, we might prejit A, and B will not end up getting inlined.

A third option is for the jit to remember if it saw any of these intrinsics during importation for a method during prejit codegen, and report this back to the VM at the end of jittting, or perhaps fail but only after inlining once all the intrinsics have been seen (or let the failure to import B as an inlinee cascade up and cause A to fail too). Then neither A or B is prejitted and at jit-time B can get inlined into A.

jkotas commented:
> We could fail prejitting (or the equivalent) when one of these intrinsics is seen in the importer either as a root method or a callee

I think this is what we should do. It is what's done for the existing SIMD methods as well.

fiigii commented:
> It would be nice to expand the tabular scheme you have introduced for the hardware intrinsics to include the non-hardware ones too (right now there's just the one, but there are two more in pending PRs, and likely more to come).
>
> In my opinion this can wait for a subsequent change where we perhaps also auto-generate the recognizer. But we should do this before opening the floodgates on the hardware intrinsic set.

@AndyAyersMS Will do when I enable the first class of hardware intrinsics, which may open the "floodgates".

fiigii commented on src/jit/importer.cpp:
```diff
@@ -3324,7 +3324,7 @@ GenTreePtr Compiler::impIntrinsic(GenTreePtr            newobjThis,
     // We disable the inlining of instrinsics for MinOpts.
     //
-    if (!mustExpand && (opts.compDbgCode || opts.MinOpts()))
+    if (!mustExpand && (opts.compDbgCode || opts.MinOpts()) && !gtIsRecursiveCall(method))
```
@jkotas I tried to specially treat recursively called intrinsics by this check, and it can avoid infinite recursion but `IsSupported` always returns `false`. Could you take a look?

jkotas commented on src/mscorlib/System.Private.CoreLib.csproj:
```diff
@@ -283,6 +285,23 @@
     <Compile Include="$(BclSourcesRoot)\System\Runtime\Intrinsics\X86\Ssse3.cs" />
   </ItemGroup>
+  <ItemGroup Condition="'$(Platform)' != 'amd64' AND '$(Platform)' != 'x86'">
+    <Compile Include="$(BclSourcesRoot)\System\Runtime\Intrinsics\X86\AesFallback.cs" />
```
Nit: Our convention is to have dot between the name and suffix for pattern like these. Also, fallback suffix maybe a bit misleading because the implementation is not actually true fallback - it just throws PNSE.

https://github.com/dotnet/corefx/blob/master/src/System.CodeDom/src/Microsoft/CSharp/CSharpCodeGenerator.PlatformNotSupported.cs is a good prior art. Call this `Aes.PlatformNotSupported.cs` ?

jkotas replied:
It maybe a good idea to submit the `*.cs` changes in separate PR to make the rest of the actual changes in this PR easier to review.

fiigii replied:
Thanks, this name makes more sense.

> It maybe a good idea to submit the *.cs changes in separate PR to make the rest of the actual changes in this PR easier to review.

Will do.

fiigii commented:
@jkotas @CarolEidt PTAL

fiigii commented:
@jkotas does the latest change look good to you? That passed the tests on debug and release build both.

jkotas commented on src/vm/methodtablebuilder.cpp:
```diff
@@ -1484,6 +1484,23 @@ MethodTableBuilder::BuildMethodTableThrowing(
+    LPCUTF8 x86className;
+    LPCUTF8 x86nameSpace;
+    isX86IntrinsicType = false;
+    HRESULT hrns = GetMDImport()->GetNameOfTypeDef(bmtInternal->pType->GetTypeDefToken(), &x86className, &x86nameSpace);
```
We typically just use `hr` in situations like this

jkotas commented on src/vm/methodtablebuilder.cpp:
```diff
@@ -1484,6 +1484,23 @@ MethodTableBuilder::BuildMethodTableThrowing(
     }
 
+#if defined(_TARGET_X86_) || defined(_TARGET_AMD64_) 
+    // All the funtions in System.Runtime.Intrinsics.X86 are hardware intrinsics.
```
This should be done for system module only like the other check. No need to do it for non-system module.

fiigii replied:
Thank you! Will fix.

jkotas commented on src/vm/methodtablebuilder.h:
```diff
@@ -2956,6 +2956,11 @@ class MethodTableBuilder
 
+#if defined(_TARGET_X86_) || defined(_TARGET_AMD64_) 
+private:
+    bool isX86IntrinsicType;
```
This should be called `isHardwareIntrinsic` and it should not be under ifdef - we will want the same bool for other platform intrinsics, and it should be in say bmtProperties to follow the style of this type.

fiigii replied:
> it should be in say bmtProperties to follow the style of this type.

Don't you mean that `isHardwareIntrinsic` should be a field of `bmtProperties` struct?

jkotas replied:
Right - it is what I meant

fiigii replied:
Thanks, will do.

fiigii commented:
@dotnet-bot test OSX10.12 x64 Checked Build and Test 
@dotnet-bot test Ubuntu x64 Checked Build and Tesst
@dotnet-bot test Windows_NT x64 Debug Build and Test
@dotnet-bot test Windows_NT x64 Release Priority 1 Build and Tes
@dotnet-bot test Windows_NT x86 Checked Build and Test

fiigii commented:
@dotnet-bot test this

AndyAyersMS commented on src/jit/importer.cpp:
```diff
@@ -3324,7 +3324,7 @@ GenTreePtr Compiler::impIntrinsic(GenTreePtr            newobjThis,
     // We disable the inlining of instrinsics for MinOpts.
     //
-    if (!mustExpand && (opts.compDbgCode || opts.MinOpts()))
+    if (!mustExpand && (opts.compDbgCode || opts.MinOpts()) && !gtIsRecursiveCall(method))
```
Can you please add a comment explaining the purpose of this extra check.

Also, while you're in the neighborhood, can you fix the misspelled word in the existing comment?

fiigii replied:
Got it, will do.

fiigii commented:
Addressed all the feedback, thank you all :mortar_board:.

jkotas commented on src/vm/methodtablebuilder.cpp:
```diff
@@ -5090,10 +5111,15 @@ MethodTableBuilder::InitNewMethodDesc(
 
-        if (hr == S_OK)
+        if (hr == S_OK
+#if defined(_TARGET_X86_) || defined(_TARGET_AMD64_)
```
Could you please delete this ifdef here? It is not necessary and it will be become a problem for enabling the HW intrinsics on more platforms later.

fiigii replied:
Done.

jkotas commented on src/vm/methodtablebuilder.cpp:
```diff
@@ -1484,6 +1484,27 @@ MethodTableBuilder::BuildMethodTableThrowing(
+    {
+        LPCUTF8 x86className;
+        LPCUTF8 x86nameSpace;
+        bmtProp->fIsHardwareIntrinsic = false;
```
This is not necessary - all bmtProp fields are zero initialized.

fiigii replied:
Done.

jkotas commented on src/vm/methodtablebuilder.cpp:
```diff
@@ -1484,6 +1484,27 @@ MethodTableBuilder::BuildMethodTableThrowing(
+#if defined(_TARGET_X86_) || defined(_TARGET_AMD64_) 
+    // All the funtions in System.Runtime.Intrinsics.X86 are hardware intrinsics.
+    // We specially treat them here to reduce the disk footprint of mscorlib.
+    if (GetModule()->IsSystem())
```
Actually, would it possible to more this check to `MethodTableBuilder::CheckForSystemTypes` ? We do fetch the name there already - it is better to do it just once.

fiigii replied:
This change makes more sense, will try.

fiigii replied:
@jkotas In `BuildMethodTableThrowing`, `CheckForSystemTypes` is called after `AllocAndInitMethodDescs();` that sets every MethodDesc to `isJitIntrinsic` or not. Would it possible to chenge these call sites order?

jkotas replied:
Hmmm, reordering these methods would likely require more elaborate refactoring - not something to do as part of this change.

I think it is ok to leave it as is. Maybe reduce the number of cases where we fetch the name by adding `&& !bmtGenerics->HasInstantiation()` condition in addition to the IsSystem condition.

fiigii replied:
Done.

fiigii commented on tests/src/Common/test_dependencies/test_dependencies.csproj:
```diff
@@ -22,6 +22,9 @@
       <Version>$(CoreFxPackageVersion)</Version>
     </PackageReference>
+    <PackageReference Include="System.Runtime.Intrinsics.X86">
+      <Version>4.5.0-preview1-25718-03</Version>
```
@mellinoe can I set a more flexible package version here?

mellinoe replied:
I think you can just use this like the one above:

`<Version>$(CoreFxPackageVersion)</Version>`

That property will get automatically updated every so often and the package will get upgraded.

fiigii replied:
Last week this property did not work, but today it works fine... Thank you!

fiigii commented:
@jkotas could you please merge this PR if it looks good?

jkotas commented:
I will merge once the CI is green. Thank you!

fiigii commented:
Thanks!

fiigii commented:
@dotnet-bot test Windows_NT x86 Checked Build and Test

fiigii commented:
@jkotas there are somebodies else waiting for this. Could you please merge?
